app/models.py

```diff
@@ -1643,6 +1643,57 @@ def serialize(self):
 
         return serialized
 
+    def serialize_with_cost_info(self):
+        serialized = self.serialize()
+        serialized["cost_details"] = {}
+        serialized["cost_in_pounds"] = "0.00"
+        serialized["is_cost_data_ready"] = True
+
+        if self.notification_type == "sms":
+            return self._get_cost_info_for_sms(serialized)
+        elif self.notification_type == "letter":
+            return self._get_cost_info_for_letter(serialized)
+
+        return serialized
+
+    def _get_cost_info_for_sms(self, serialized):
+        serialized["cost_details"]["sms_fragments"] = self.billable_units
+        serialized["cost_details"]["rate_multiplier"] = self.rate_multiplier
+        sms_rate = self._get_sms_rate()
+        serialized["cost_details"]["rate"] = str(sms_rate)
+        serialized["cost_in_pounds"] = str(self.billable_units * self.rate_multiplier * sms_rate)
+
+        return serialized
+
+    def _get_cost_info_for_letter(self, serialized):
+        if self.status == "pending-virus-check":  # are there any other not-ready states?
+            serialized["is_cost_data_ready"] = False
+            serialized["cost_details"] = {}
+            serialized["cost_in_pounds"] = None
+        else:
+            serialized["cost_details"]["sheets_of_paper"] = self.billable_units
```

Comment:
So for letters we will expose `serialized["cost_details"]["sheets_of_paper"]` but for text messages there won't be a `sheets_of_paper` key in `cost_details`.

Is this OK?

Obviously sheets of paper isn't relevant for text messages, but does that mean we should not include the key or we should include the key but set the value as null?

Our precedent for the top level keys in a notification is to return all the keys possible, with them set as null if they aren't appropriate (see https://github.com/alphagov/notifications-api/blob/main/app/models.py#L1601).

I don't know why we do that though... One to find out the reason for I think. It might be that that pattern is 'better API design' according to the hive mind. Maybe it is because some API users might struggle to process responses if they don't have all the keys all the time (I'm thinking maybe .net or java here). Maybe it is because if we don't always include the keys then our users have to do logic like

`if key exists and get value of key` rather than just `get value of key` everytime they want to access a key in `cost_details`.

Reply:
Ooh, good point - I will try to find out if my current approach works with different languages (I hope it does).

As to users having to do extra work to see if key doesn't exist - I'm afraid for cost info they might need to do some checks anyway:
- they will need to check the value for `"is_cost_data_ready"` - if this is False, the  "cost_in_pounds" is None instead of a string with number inside. This is on purpose - I didn't want users to assume the cost is 0.00 for notifications that are not ready.
- they will have to check if the notification is a letter or an sms to get the correct cost information (but they should know this beforehand as they are the ones sending the GET request, right?)

Anyway, that's a really interesting point and I will check about .net and java and about any API design guidelines on this 🙌🏼

Reply:
I struggled to find blog posts on this, so I looked into the Paypal API again, to see if they have examples either way. Since they are such a huge company, their users use all sorts of languages.

I saw they for example have a `"supplementary_data"` key, where the value is a dictionary that can have a variety of key-value pairs. Look for it on this page if you're interested: https://developer.paypal.com/docs/api/payments/v2/

We already also return dictionaries with variety of key-value pairs inside already. We do this for personalisation - as some templates have no placeholders, and others have some and they are always different. Here is an example:
https://docs.notifications.service.gov.uk/java.html#get-a-template-by-id-response

So since the top level key is always there, and the value is always a dict, I vote that this is good enough, and actually imo neater than the alternative, and less likely to result in users mistakenly counting the cost of notification as 0 if they confuse the notification type.

Reply:
i am in favour of not including keys for fields that do not make semantic sense

how we handle this in client libraries of strongly typed languages is a separate issue, but probably solved by polymorphism. eg java client could return a `CostData` object which is of type `LetterCostData` or `SMSCostData`

Reply:
i'm interested in whether `cost_in_pounds` should be nested within `cost_data`. i like the idea of being able to retrieve the entire cost data from the json blob in one line, and the actual value _is_ data about the cost

Reply:
I placed `cost_in_pounds` at the top level to make it's easily accessible to the users. My assumption here is that most users will be only interested in the cost itself, while `cost_details` is more like supplementary information in case they need to do some deeper digging.

I would prefer to receive it like that if I were a user.

Reply:
curious minor question: why explicitly `cost_in_pounds`, do we calculate in any other currency?